js-packages/@prestojs/rest/src/CursorPaginator.ts

```diff
@@ -0,0 +1,161 @@
+import { EndpointExecuteOptions } from './Endpoint';
+import Paginator from './Paginator';
+
+export type CursorPaginationState = {
+    pageSize?: number;
+    cursor?: string;
+};
+
+export type InternalCursorPaginatorState = {
+    nextCursor?: string | null;
+    previousCursor?: string | null;
+};
+
+export default class CursorPaginator extends Paginator<
+    CursorPaginationState,
+    InternalCursorPaginatorState
+> {
+    get nextCursor(): string | null {
+        return this.internalState.nextCursor ?? null;
+    }
+
+    get previousCursor(): string | null {
+        return this.internalState.previousCursor ?? null;
+    }
+
+    get cursor(): string | null | undefined {
+        return this.currentState.cursor;
+    }
+
+    get pageSize(): number | null | undefined {
+        return this.currentState.pageSize;
+    }
+
+    /**
+     * Return the state for the specified page size
+     *
+     * Does not transition state. To transition state call `setPageSize` instead.
+     */
+    pageSizeState(pageSize: number | null): CursorPaginationState {
+        if (pageSize != null && pageSize < 1) {
+            throw new Error(`Invalid pageSize ${pageSize} - should be >= 1`);
+        }
+
+        // If page size changes any next/previous cursors are no invalid
+        if (this.currentState.pageSize !== pageSize) {
+            this.setInternalState({ nextCursor: null, previousCursor: null });
+        }
+        const nextState = { ...this.currentState, pageSize };
+        if (pageSize == null) {
+            delete nextState.pageSize;
+        }
+        return nextState as CursorPaginationState;
+    }
+
+    setPageSize(pageSize: null | number): void {
+        this.setCurrentState(this.pageSizeState(pageSize));
+    }
+
+    /**
+     * Return the state for the first page
+     *
+     * Does not transition state. To transition state call `first` instead.
+     */
+    firstState(): CursorPaginationState {
+        // first page is equivalent to not specifying cursor
+        const nextState: CursorPaginationState = { ...this.currentState };
+        delete nextState.cursor;
+        return nextState;
+    }
+
+    /**
+     * Go to the first page.
+     */
+    first(): void {
+        this.setCurrentState(this.firstState());
+    }
+
+    /**
+     * Return the state for the first page. If the next page isn't yet known (eg. results
+     * haven't yet been returned) then null will be returned.
+     *
+     * Does not transition state. To transition state call `next` instead.
+     */
+    nextState(): CursorPaginationState | null {
+        if (this.nextCursor == null) {
+            return null;
```

Comment:
What will happen if a consumer doesn't do a check and just calls setState on the output from this and that is null?

Is it better to throw an exception here? (Am distinguishing this from individual parts of the state, eg `total` because those are used to compose the state but you wouldn't call setState() directly with their output)

Reply:
The are the places I previously was throwing exceptions from that you objected to that I changed to return null... eg. so people can call them in generic components

Reply:
The distinction is between setters/actions & getters:
- if you say "take me to the last page" but there's no way to do that then you need to communicate that back to the caller: either a success/fail return value or throw an exception
- the same logic applies to a getter: you can either return a failure value (`null`) or throw an exception. The difference between getters & setters is that catching an exception in JSX vs checking for a null return value is very clumsy.

Something like:

`{paginator.lastPage() && <button onClick={paginator.gotoLastPage()}>Last</button>}`

vs 

`{(() => { try { return <button >Last</button>} catch (err) { if (err.type instanceof WhateverException) throw err; return null; })()}`


Although this doesn't really apply if there is a standard helper to ignore exception(s), something like:
`{ignoreException(WhateverException, <button >Last</button>)}`

Reply:
I understand all those points but don't understand what you want especially given previous comments

Can you clarify?

You've commented on `nextState` here - eg. give me the state for the 'next' page. The whole purpose of this was so you could do things like generate a URL with it which was the cases you were talking about for _not_ throwing exceptions.

Reply:
Note to self: check what currently throws; it should allow things like running of the end with next but stuff that doesn't make sense throw for (eg. going to last() on a cursor paginator)